mava/systems/q_learning/anakin/rec_qmix.py

```diff
@@ -0,0 +1,697 @@
+
+        next_obs = next_timestep.observation
+        # Make compatible with network input and transition storage in next step
+        next_terminal = (1 - next_timestep.discount[..., 0, jnp.newaxis]).astype(bool)
+        next_term_or_trunc = next_timestep.last()[..., jnp.newaxis]
+
+        new_act_state = ActionState(
+            next_action_selection_state,
+            next_env_state,
+            next_buffer_state,
+            next_obs,
+            next_terminal,
+            next_term_or_trunc,
+        )
+
+        return new_act_state, next_timestep.extras["episode_metrics"]
+
+    def prep_inputs_to_scannedrnn(obs: Observation, term_or_trunc: chex.Array) -> chex.Array:
+        """Prepares the inputs to the RNN network for either getting q values or the
+        eps-greedy distribution.
+
+        Mostly swaps leading axes because the replay buffer outputs (B, T, ... )
+        and the RNN takes in (T, B, ...).
+        """
+        hidden_state = ScannedRNN.initialize_carry(
+            (cfg.system.sample_batch_size, obs.agents_view.shape[2]), cfg.network.hidden_state_dim
+        )
+        # the rb outputs (B, T, ... ) the RNN takes in (T, B, ...)
+        obs = switch_leading_axes(obs)  # (B, T) -> (T, B)
+        term_or_trunc = switch_leading_axes(term_or_trunc)  # (B, T) -> (T, B)
+        obs_term_or_trunc = (obs, term_or_trunc)
+
+        return hidden_state, obs_term_or_trunc
+
+    def q_loss_fn(
+        online_params: FrozenVariableDict,
+        obs: Array,
+        term_or_trunc: Array,
+        action: Array,
+        target: Array,
+    ) -> Tuple[Array, Metrics]:
+        """The portion of the calculation to grad, namely online apply and mse with target."""
+        q_online_params, online_mixer_params = online_params
+
+        # Axes switched to scan over time
+        hidden_state, obs_term_or_trunc = prep_inputs_to_scannedrnn(obs, term_or_trunc)
+
+        # Get online q values of all actions
+        _, q_online = q_net.apply(
+            q_online_params, hidden_state, obs_term_or_trunc, method="get_q_values"
+        )
+        q_online = switch_leading_axes(q_online)  # (T, B, ...) -> (B, T, ...)
+        # Get the q values of the taken actions and remove extra dim
+        q_online = jnp.squeeze(
+            jnp.take_along_axis(q_online, action[..., jnp.newaxis], axis=-1), axis=-1
+        )
+
+        q_online = mixer.apply(
+            online_mixer_params, q_online, obs.global_state[:, :, 0, ...]
+        )  # B,T,A,... -> B,T,1,...
+        # NOTE: States are replicated over agents so we take only take first one
+
+        q_loss = jnp.mean((q_online - target) ** 2)
+
+        loss_info = {
+            "q_loss": q_loss,
+            "mean_q": jnp.mean(q_online),
+            "max_q_error": jnp.max(jnp.abs(q_online - target) ** 2),
+            "min_q_error": jnp.min(jnp.abs(q_online - target) ** 2),
+            "mean_target": jnp.mean(target),
```

Comment:
Doing this calculation twice, can you store it in an error variable and then min/max that :pray: